qutebrowser/browser/greasemonkey.py

```diff
@@ -171,23 +194,110 @@ def load_scripts(self):
-                                             "document-end"
-                                             .format(script_path))
-                    # Default as per
-                    # https://wiki.greasespot.net/Metadata_Block#.40run-at
-                    self._run_end.append(script)
-                log.greasemonkey.debug("Loaded script: {}".format(script.name))
+                    self._add_script(script)
+
         self.scripts_reloaded.emit()
 
+    def _add_script(self, script):
+        if script.run_at == 'document-start':
+            self._run_start.append(script)
+        elif script.run_at == 'document-end':
+            self._run_end.append(script)
+        elif script.run_at == 'document-idle':
+            self._run_idle.append(script)
+        else:
+            log.greasemonkey.warning("Script {} has invalid run-at "
+                                     "defined, defaulting to "
+                                     "document-end"
+                                     .format(script.name))
+            # Default as per
+            # https://wiki.greasespot.net/Metadata_Block#.40run-at
+            self._run_end.append(script)
+        log.greasemonkey.debug("Loaded script: {}".format(script.name))
+
+    def _required_url_to_file_path(self, url):
+        # Save to a more readable name
+        # cf https://stackoverflow.com/questions/295135/turn-a-string-into-a-valid-filename
+        # replace everything except hyphen, word characters, underscore and dot
+        # with and underscore.
+        name = re.sub(r'(?u)[^-\w.]', '_', url)
+        requires_dir = os.path.join(_scripts_dir(), 'requires')
+        if not os.path.exists(requires_dir):
+            os.mkdir(requires_dir)
+        return os.path.join(requires_dir, name)
+
+    def _on_required_download_finished(self, script, download=None,
+                                       quiet=False):
+        if download:
+            self._in_progress_dls.remove(download)
+
+        # See if we are still waiting on any required scripts for this one
+        for dl in self._in_progress_dls:
+            if dl.requested_url in script.requires:
+                log.greasemonkey.debug(("got dl for script {} but some "
+                                        "requirements still pending")
+                                       .format(script.name))
+                return
+
+        # Need to add the required scripts to the iife now
+        for url in reversed(script.requires):
+            target_path = self._required_url_to_file_path(url)
+            log.greasemonkey.debug(("Adding required script for {} to iife: "
+                                    "{}").format(script.name, url))
+            with open(target_path, encoding='utf8') as f:
+                script.add_required_script(f.read())
+
+        self._add_script(script)
+        if not quiet:
+            self.scripts_reloaded.emit()
+
+    def _start_pending_downloads(self, _window):
```

Comment:
No need to take a `_window` parameter here - signals can be connected to slots with fewer elements. You should decorate it with `@pyqtSlot()` though.